extensions/stackdriver/common/metrics.cc

```diff
@@ -0,0 +1,62 @@
+/* Copyright 2020 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/stackdriver/common/metrics.h"
+
+#ifdef NULL_PLUGIN
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+#endif
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Common {
+
+// NOTE: every call to this function will alloc a new metric object. Though this
+// object is the same in all calls, it cannot be extracted out to a global
+// variable. The metric object caches all fully resolved metrics and won't make
+// define metrics abi call when the same resolved metrics are defined. In Null
+// VM case, since global objects are shared between root contexts in base VM and
+// in thread local VM, and the same metrics definition code path will excercised
+// in both VMs (we don't really have a way to distinguish base VM and thread
+// local VM), the cache will make thread local root context miss metric
+// definition at host side. In real Wasm VM, this can be extracted to a global
+// var.
```

Comment:
I'm not sure I understand. Is there a performance cost and difference between null VM and v8? I thought the cache is on the host side.

Reply:
Performance is not a concern, these function will only be called during initialization. Essentially what causes problem is the asymmetric ownership of metric object: Inside VM metric is not attached to a specific context, but it attaches to a VM/root context in host. This should be fine in a real wasm VM.

I originally make it a static global var `static Metric export_call(...)`, and inside the metric object, there is a map to keep currently fully resolved metrics: https://github.com/envoyproxy/envoy-wasm/blob/f6d0282f7ab42dfd215ae59112edbab3e70f2d3b/api/wasm/cpp/proxy_wasm_api.h#L831 What happened is that in nullvm case, thread local VM and main thread base VM share the same global metric object, [the metric id map](https://github.com/envoyproxy/envoy-wasm/blob/f6d0282f7ab42dfd215ae59112edbab3e70f2d3b/api/wasm/cpp/proxy_wasm_api.h#L831) is filled during base VM root context creation and [blocks](https://github.com/envoyproxy/envoy-wasm/blob/f6d0282f7ab42dfd215ae59112edbab3e70f2d3b/api/wasm/cpp/proxy_wasm_api.h#L917) proxy_define_metric ABI call in thread local VM root context, where metric is really incremented. It is not going to be a problem for real wasm VM because of memory isolation.

Hope this makes sense, let me see how can i make the comment more clear.

Reply:
I updated the comment. Hope it is clearer than the original one, or at least not making it more confusing. :p